### PR TITLE
Update LSTM Attention Model to use tf.contrib.seq2seq.AttentionWrapper

### PR DESCRIPTION
**Pull request by kolloldas**
The LSTM Attention model was updated:
1. Replaced old attention model with `AttentionWrapper` based code
2. *HParams* added for Bahdanau Type Attention and Luong Type Attention
3. *HParams* added for multi-head attention
4. Set weight decay to 0 to fix NaN training loss issues

I did a quick comparison with the variants on the *translate_ende_wmt_bpe32k* task. 
After 2000 train steps:

| Model                                 | Parameters | Validation Loss  | BLEU | Accuracy |
| ----------------------------- | ------------- | -----------------|-------|-----------|
| LSTM Bahdanau Attention  | 14867456 | 6.25 | 0.2351 | 11.73% |
| LSTM Luong Attention        | 14850944  | 6.27 | 0.2319 | 11.72% |
| LSTM Bahdanau Attention Multi (4)  | 15228416 | 6.18 | 0.2370 | 11.99% |
| LSTM Luong Attention Multi (4)       | 15211520 | 6.16 | 0.2358 | 12.27% |
| Transformer (Base Single GPU)        | 63103488 | 5.95 | 0.1223 | 14.22% |

Please review, thanks!

### CODE REVIEW

**Comment by martinpopel** [comment]
Is there any explanation for the disagreement between BLEU and Accuracy (or Validation loss)? Transformer has the best accuracy, but the worst BLEU according to your table.

**Comment by kolloldas** [comment]
@martinpopel there appears to be some problem with the way BLEU is computed by t2t, almost as if lower is better! Here are the results after 5000 train steps. I've now included Rouge-L scores as well:

| Model                                 | Parameters | Validation Loss  | BLEU | Accuracy | Rouge-L |
| ----------------------------- | ------------- | -----------------|-------|-----------|---------|
| LSTM Bahdanau Attention  | 14867456 | 5.76 | 0.1818 | 14.58% | 0.169 |
| LSTM Luong Attention        | 14850944  | 5.75 | 0.1965 | 14.52% | 0.168 |
| LSTM Bahdanau Attention Multi (4)  | 15228416 | 5.67 | 0.1394 | 14.84% | 0.175 |
| LSTM Luong Attention Multi (4)       | 15211520 | 5.64 | 0.1311 | 15.47% | 0.179 |
| Transformer (Base Single GPU)        | 63103488 | 4.73 | 0.0636 | 24.02% | 0.279 |

As you can see the differences between transformer and the rest are more pronounced.

**Comment by vince62s** [comment]
I think you need to wait 50k or 100k to get meaningful values.
But a few questions:
What learning rate / decay did you take for LSTM ?
Do I read well saying you took 2x128 as default for layers / hidden size ?

**Comment by kolloldas** [comment]
Yes that's true, 5K steps are not enough, but it was more of a sanity check.
I kept the default hyperparameters in T2T:
learning_rate = 0.1
num_layers = 2
hidden_size = 128

**Comment by kolloldas** [comment]
Great! 
Should be able to close #113 #205 #375 #378

**Comment by martinpopel** [comment]
@kolloldas Do the approx_bleu values after 100k steps look reasonable?

I've noticed the approx_bleu curve always starts from quite high values at 0 steps (sometimes even 0.8), then drops (e.g. to 0.05 at 10k steps) and then starts growing and only from that point it correlates with the real BLEU.
Does anyone know why?
Is it possible that approx_bleu is computed in a way that when predicting the n-th (subword) token, tokens 1..n-1 *from the reference translation* are known (and backpropagated), so the network first tries to repeat the last token (and shifting words by one has just a minor effect on BLEU).

**Comment by kolloldas** [comment]
I'm yet to run this task for 100K steps but I think BLEU is initially high because the model makes very short predictions at the beginning. The BLEU drops as predictions get longer (but are mostly gibberish) and then finally begins to pick up as predictions become more meaningful..

**Comment by martinpopel** [comment]
No. BLEU contains a *brevity penalty* term, so too short translations get very low BLEU. I think the reason is what I have described above, ie. approx_bleu by default is not computed autoregressively, but using previous gold words as input for the decoder. You can use `--eval_run_autoregressive`, but this seems to be currently buggy (#407).
